// minecraft-server-ping/dist/PacketDecoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketDecoder = void 0;
const bigint_buffer_1 = require("bigint-buffer");
const stream_1 = require("stream");
const varint_1 = require("varint");
var PackageType;
(function (PackageType) {
    PackageType[PackageType["HANDSHAKE"] = 0] = "HANDSHAKE";
    PackageType[PackageType["PING"] = 1] = "PING";
})(PackageType || (PackageType = {}));
class PacketDecoder extends stream_1.Writable {
    constructor(options) {
        super(options);
        this.buffer = Buffer.alloc(0);
    }
    oncePromise(event) {
        return new Promise((resolve) => {
            this.once(event, resolve);
        });
    }
    async _write(chunk, encoding, callback) {
        const { getPayload, decodeHandshake, decodePong } = this;
        if (!this.packetInfo) {
            this.packetInfo = this.decodeHeader(chunk);
        }
        this.buffer = Buffer.concat([this.buffer, chunk]);
        if (this.packetInfo) {
            if (this.buffer.length < this.packetInfo.length) {
                // do we still need to read more?
                return callback();
            }
            if (this.buffer.length > this.packetInfo.length) {
                throw new Error('we did overrun expected data size!');
            }
        }
        try {
            if (this.packetInfo) {
                switch (this.packetInfo.id) {
                    case PackageType.HANDSHAKE: {
                        this.emit('handshake', decodeHandshake(getPayload(this.packetInfo, this.buffer)));
                        break;
                    }
                    case PackageType.PING: {
                        this.emit('pong', decodePong(getPayload(this.packetInfo, this.buffer)));
                        break;
                    }
                    default:
                        this.emit('error', new Error('Unknown packet id: ' + this.packetInfo.id));
                }
            }
            this.buffer = Buffer.alloc(0);
            this.packetInfo = undefined;
        }
        catch (err) {
            this.emit('error', err);
        }
        callback();
    }
    decodeHeader(buffer) {
        const length = varint_1.decode(buffer);
        return {
            id: buffer.readUInt8(varint_1.encodingLength(length)),
            length: length + varint_1.encodingLength(length),
            offset: varint_1.encodingLength(length) + 3,
        };
    }
    getPayload(header, data) {
        return data.slice(header.offset, data.length);
    }
    decodeHandshake(data) {
        return JSON.parse(data.toString());
    }
    decodePong(data) {
        return Number(BigInt(Date.now()) - bigint_buffer_1.toBigIntBE(data));
    }
}
exports.PacketDecoder = PacketDecoder;
